zdm/analyze_cube.py

```diff
@@ -0,0 +1,477 @@
+    """ Method to perform simple Bayesian analysis
+    on the Log-likelihood cube
+
+    Args:
+        lls (np.ndarray): Log-likelood cube
+        plls (np.ndarray, optional): Log-likelihood cube corrected for priors (e.g. alpha). Defaults to None.
+        pklfile (str, optional): If given, write
+            the output to this pickle file. Defaults to None.
+
+    Returns:
+        tuple: uvals,vectors,wvectors
+            lists of np.ndarray's of LL analysis
+            One item per parameter in the cube
+    """
+    NDIMS= len(lls.shape)
+            
+    origlls=lls
+    if plls is None:
+        plls = lls
+    uvals=[]
+    
+    for i in np.arange(NDIMS):
+        unique = np.arange(lls.shape[i])
+        uvals.append(unique)
+
+    # we now have a list of unique values for each dimension
+    vectors=[] # this will contain the best values for 1d plots
+    wvectors=[] # holds same as above, but including spectral penalty factor from ASKAP obs
+    
+    # loop over the DIMS
+    for i in np.arange(NDIMS):
+        
+        # does 1D values
+        vector=np.zeros([len(uvals[i])])
+        wvector=np.zeros([len(uvals[i])])
+
+        # selects for lls a subset corresponding only to that particular value of a variables
+        for iv, ivv in enumerate(uvals[i]):
+            big_slice = [slice(None,None,None)]*NDIMS
+            # Construct the slice
+            big_slice[i] = ivv
+            #set1=np.where(data[:,idim]==ivv) #selects for a set of values
+            #lls=data[set1,llindex]
+            lls=origlls[tuple(big_slice)].flatten()
+            wlls=plls[tuple(big_slice)].flatten()
+            
+            # selects all fits that are close to the peak (i.e. percentage within 0.1%)
+            try:
+                themax=np.nanmax(lls)
+            except:
+                # all nans, probability =0. Easy!
+                vector[iv]=0.
+                wvector[iv]=0.
+                continue
+            
+            wthemax=np.nanmax(wlls)
+            OKlls=np.isfinite(lls) & (lls > themax-3)
+            OKwlls=np.isfinite(lls) & (wlls > wthemax-3)
+            
+            vector[iv]=np.sum(10**lls[OKlls])
+            wvector[iv]=np.sum(10**wlls[OKwlls])
+        vector *= 1./np.sum(vector)
+        wvector *= 1./np.sum(wvector)	
+        vectors.append(vector)
+        wvectors.append(wvector)
+    
+    # Pickle?
+    if pklfile is not None:
+        with open(pklfile, 'wb') as output:
+            pickle.dump(uvals, output, pickle.HIGHEST_PROTOCOL)
+            pickle.dump(vectors, output, pickle.HIGHEST_PROTOCOL)
+            pickle.dump(wvectors, output, pickle.HIGHEST_PROTOCOL)
+        
+    # result is just the total probability, normalised to unit, when summed over the parameter space
+    # technically needs to be divided by the x-increment in bins.
+    return uvals,vectors,wvectors
+
+
+def do_single_plots(uvals,vectors,wvectors,names,tag=None, fig_exten='.png',
```

Comment:
FYI, a lot of this routine is affected by fluff I put in to make the plots more publication ready, but maybe leave for now? Another thing to do might be to extract all the useful info, save it, and then do the plotting in another routine? Anyway, only possible to-dos.

Reply:
yeah, let's leave for now.